lib/parser/livesplit_core_parser.rb

```diff
@@ -62,95 +75,105 @@ def self.parse(run, fast: false)
-        split.indexed_history  = {}
-        split.realtime_history = []
-        until (history_element = history_iterator.next).nil?
-          history_element_id   = history_element.index
-          history_element_time = history_element.time
-          split.indexed_history[history_element_id] = {
-            gametime: history_element_time.game_time.try(:total_seconds) || 0,
-            realtime: history_element_time.real_time.try(:total_seconds) || 0
-          }
-
-          split.realtime_history << history_element_time.real_time.try(:total_seconds) || 0
-        end
-        history_iterator.dispose
-      end
+      split[:gametime_end_ms] = (segment.personal_best_split_time.game_time.try(:total_seconds) || 0) * 1000
+      split[:gametime_duration_ms] = [0, split[:gametime_end_ms] - run_object[:gametime_duration_ms]].max
+      split[:gametime_start_ms] = split[:gametime_end_ms] - split[:gametime_duration_ms]
 
-      split.gametime_end = segment.personal_best_split_time.game_time.try(:total_seconds) || 0
-      split.gametime_duration = [0, split.gametime_end - run_object[:gametime_time]].max
-      split.gametime_start = split.gametime_end - split.gametime_duration
+      split[:gametime_best_ms] = (segment.best_segment_time.game_time.try(:total_seconds) || 0) * 1000
+      split[:gametime_skipped] = split[:gametime_duration_ms].zero?
 
-      split.gametime_best = segment.best_segment_time.game_time.try(:total_seconds) || 0
-      split.gametime_skipped = split.gametime_duration.zero?
+      if split[:gametime_duration_ms].positive?
+        split[:gametime_gold] = true if split[:gametime_duration_ms].round(5) <= split[:gametime_best_ms].try(:round, 5)
+      end
 
-      if split.gametime_duration.positive? && split.gametime_duration.round(5) <= split.gametime_best.try(:round, 5)
-        split.gametime_gold = true
-      else
-        split.gametime_gold = false
+      history_iterator = segment.segment_history.iter
+      until (history_element = history_iterator.next).nil?
+        history_element_time = history_element.time
+        split[:history] << {
+          attempt_number: history_element.index,
+          gametime_duration_ms: (history_element_time.game_time.try(:total_seconds) || 0) * 1000,
+          realtime_duration_ms: (history_element_time.real_time.try(:total_seconds) || 0) * 1000
+        }
       end
+      history_iterator.dispose
 
-      run_object[:realtime_time] += split.realtime_duration if split.realtime_duration.present?
-      run_object[:gametime_time] += split.gametime_duration if split.gametime_duration.present?
+      run_object[:realtime_duration_ms] += split[:realtime_duration_ms] if split[:realtime_duration_ms].present?
+      run_object[:gametime_duration_ms] += split[:gametime_duration_ms] if split[:gametime_duration_ms].present?
       run_object[:splits] << split
     end
 
     run.dispose
+    traverse(run_object)
     run_object
   end
+
+  def self.traverse(obj)
```

Comment:
Does this modify `obj` or create a new one and return it? I think you have to do something special in Ruby to modify an argument and have its changes persist in the caller.

If you do, I'd also add a `!` to the end of this method name to communicate that it does that.

Reply:
It does modify the original hash, just forgot about the `!` and `?` stuff that ruby does.  I don't know about doing anything special though, since what it does currently modified it since I was testing it out in the console.